src/PgRvn.Server/PowerBI/PBIFetchQuery.cs

```diff
@@ -2,210 +2,363 @@
                 {
-                    var (sourceColumn, replaceInput, replaceText) = value;
-                    projection += $"\"{columnName}\": {alias}.{sourceColumn}.toString().replace(\"{replaceInput}\", \"{replaceText}\")";
+                    destColumnVal = val;
                 }
-                //else if (selectFieldValues != null) // TODO: Support this, i don't think it works as it is
-                //{
-                //    projection += $"\"{columnName}\": ({selectFieldValues.Captures[i].Value})";
-                //}
                 else
                 {
-                    projection += $"\"{columnName}\": {alias}.{columnName}";
+                    // Don't create new fields without a source when projectionFields is 
+                    // already populated from a previous SQL/RQL layer
+                    if (alreadyPopulatedByPreviousLayer)
+                    {
+                        continue;
+                    }
+
+                    destColumnVal = GenerateColumnValue(destColumn, alias);
                 }
 
-                projection += ", ";
+                if (replace[i].Value.Length != 0)
+                {
+                    destColumnVal = $"({destColumnVal}).toString()" +
+                        $".replace(\"{replaceInputs[replaceIndex].Value}\", \"{replaceTexts[replaceIndex].Value}\")";
+                    replaceIndex++;
+                }
+
+                projectionFields[destColumn] = destColumnVal;
+            }
+        }
+
+        private static string GenerateColumnValue(string column, string alias)
+        {
+            var val = $"{alias}[\"{column}\"]";
+
+            // TODO: Support replacing json()
+            // TODO: Won't work on queries with include because in RqlQuery.Execute included documents' ids are put into id()
+            if (column.Equals("id()", StringComparison.OrdinalIgnoreCase))
+            {
+                val = $"{alias}[\"@metadata\"][\"@id\"]";
+            }
+
+            return val;
+        }
+
+        private static string GenerateProjectedRql(Match rqlMatch, List<KeyValuePair<string, string>> projectionFields, ICollection<Match> matches)
+        {
+            var rql = rqlMatch.Value;
+
+            var collection = rqlMatch.Groups["collection"];
+            var alias = rqlMatch.Groups["alias"];
+            var where = rqlMatch.Groups["where"];
+            var select = rqlMatch.Groups["select"];
+
+            var aliasIndex = alias.Success ? (alias.Index + alias.Length) : (collection.Index + collection.Length);
+            var aliasLength = alias.Length;
+
+            var whereIndex = where.Success ? where.Index : (aliasIndex + aliasLength);
+            var whereLength = where.Length;
+
+            var selectIndex = select.Success ? select.Index : (whereIndex + whereLength);
+
+            // Insert alias clause if doesn't exist
+            if (!alias.Success)
```

Comment:
Minor - as a convention we always use `if (alias.Success == false)` instead of `if (!alias.Success)`. It's easier to read it

Reply:
Done